File: src/main/java/com/mbayou/kick4k/KickConfiguration.java
package com.mbayou.kick4k;

import com.mbayou.kick4k.authorization.RefreshTokenStore;

public final class KickConfiguration {
    private final RefreshTokenStore tokenStore;

    private final String clientId;
    private final String clientSecret;
    private final String redirectUri;

    private final String oAuthHost;
    private final String authorizationEndpoint;
    private final String tokenEndpoint;
    private final String baseUrl;
    private final String categories;
    private final String categoriesId;
    private final String tokenIntrospect;
    private final String users;
    private final String channels;
    private final String chat;
    private final String moderation;
    private final String livestreams;
    private final String livestreamsStats;
    private final String publicKey;
    private final String events;

    private KickConfiguration(Builder builder) {
        this.tokenStore = builder.tokenStore;
        this.clientId = builder.clientId;
        this.clientSecret = builder.clientSecret;
        this.redirectUri = builder.redirectUri;
        this.oAuthHost = builder.oAuthHost;
        this.authorizationEndpoint = builder.authorizationEndpoint;
        this.tokenEndpoint = builder.tokenEndpoint;
        this.baseUrl = builder.baseUrl;
        this.categories = builder.categories;
        this.categoriesId = builder.categoriesId;
        this.tokenIntrospect = builder.tokenIntrospect;
        this.users = builder.users;
        this.channels = builder.channels;
        this.chat = builder.chat;
        this.moderation = builder.moderation;
        this.livestreams = builder.livestreams;
        this.livestreamsStats = builder.livestreamsStats;
        this.publicKey = builder.publicKey;
        this.events = builder.events;
    }

    public static Builder builder() {
        return new Builder();
    }

    public RefreshTokenStore getTokenStore() {
        return tokenStore;
    }

    public String getClientId() {
        return clientId;
    }

    public String getClientSecret() {
        return clientSecret;
    }

    public String getRedirectUri() {
        return redirectUri;
    }

    public String getOAuthHost() {
        return oAuthHost;
    }

    public String getAuthorizationEndpoint() {
        return authorizationEndpoint;
    }

    public String getTokenEndpoint() {
        return tokenEndpoint;
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    public String getCategories() {
        return categories;
    }

    public String getCategoriesId() {
        return categoriesId;
    }

    public String getTokenIntrospect() {
        return tokenIntrospect;
    }

    public String getUsers() {
        return users;
    }

    public String getChannels() {
        return channels;
    }

    public String getChat() {
        return chat;
    }

    public String getModeration() {
        return moderation;
    }

    public String getLivestreams() {
        return livestreams;
    }

    public String getLivestreamsStats() {
        return livestreamsStats;
    }

    public String getPublicKey() {
        return publicKey;
    }

    public String getEvents() {
        return events;
    }

    public static final class Builder {
        private RefreshTokenStore tokenStore;
        private String clientId;
        private String clientSecret;
        private String redirectUri;
        private String oAuthHost = "https://id.kick.com";
        private String authorizationEndpoint = "/oauth/authorize";
        private String tokenEndpoint = "/oauth/token";
        private String baseUrl = "https://api.kick.com/public/v1";
        private String categories = "/categories";
        private String categoriesId = "/categories/{id}";
        private String tokenIntrospect = "/token/introspect";
        private String users = "/users";
        private String channels = "/channels";
        private String chat = "/chat";
        private String moderation = "/moderation/bans";
        private String livestreams = "/livestreams";
        private String livestreamsStats = "/livestreams/stats";
        private String publicKey = "/public-key";
        private String events = "/events/subscriptions";

        public Builder tokenStore(RefreshTokenStore tokenStore) {
            this.tokenStore = tokenStore;
            return this;
        }

        public Builder clientId(String clientId) {
            this.clientId = clientId;
            return this;
        }

        public Builder clientSecret(String clientSecret) {
            this.clientSecret = clientSecret;
            return this;
        }

        public Builder redirectUri(String redirectUri) {
            this.redirectUri = redirectUri;
            return this;
        }

        public Builder oAuthHost(String oAuthHost) {
            this.oAuthHost = oAuthHost;
            return this;
        }

        public Builder authorizationEndpoint(String authorizationEndpoint) {
            this.authorizationEndpoint = authorizationEndpoint;
            return this;
        }

        public Builder tokenEndpoint(String tokenEndpoint) {
            this.tokenEndpoint = tokenEndpoint;
            return this;
        }

        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        public Builder categories(String categories) {
            this.categories = categories;
            return this;
        }

        public Builder categoriesId(String categoriesId) {
            this.categoriesId = categoriesId;
            return this;
        }

        public Builder tokenIntrospect(String tokenIntrospect) {
            this.tokenIntrospect = tokenIntrospect;
            return this;
        }

        public Builder users(String users) {
            this.users = users;
            return this;
        }

        public Builder channels(String channels) {
            this.channels = channels;
            return this;
        }

        public Builder chat(String chat) {
            this.chat = chat;
            return this;
        }

        public Builder moderation(String moderation) {
            this.moderation = moderation;
            return this;
        }

        public Builder livestreams(String livestreams) {
            this.livestreams = livestreams;
            return this;
        }

        public Builder livestreamsStats(String livestreamsStats) {
            this.livestreamsStats = livestreamsStats;
            return this;
        }

        public Builder publicKey(String publicKey) {
            this.publicKey = publicKey;
            return this;
        }

        public Builder events(String events) {
            this.events = events;
            return this;
        }

        public KickConfiguration build() {
            if (tokenStore == null) {
                throw new IllegalStateException("TokenStore is required");
            }

            if (clientId == null) {
                throw new IllegalStateException("ClientId is required");
            }

            if (clientSecret == null) {
                throw new IllegalStateException("ClientSecret is required");
            }

            if (redirectUri == null) {
                throw new IllegalStateException("RedirectUri is required");
            }

            return new KickConfiguration(this);
        }
    }
}

File: src/main/kotlin/com/mbayou/kick4k/KickConfiguration.kt
package com.mbayou.kick4k

import com.mbayou.kick4k.authorization.RefreshTokenStore

data class KickConfiguration(
    val tokenStore: RefreshTokenStore,
    val clientId: String,
    val clientSecret: String,
    val redirectUri: String,
    val oAuthHost: String,
    val authorizationEndpoint: String,
    val tokenEndpoint: String,
    val baseUrl: String,
    val categories: String,
    val categoriesId: String,
    val tokenIntrospect: String,
    val users: String,
    val channels: String,
    val chat: String,
    val moderation: String,
    val livestreams: String,
    val livestreamsStats: String,
    val publicKey: String,
    val events: String,
) {
    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }

    class Builder {
        private var tokenStore: RefreshTokenStore? = null
        private var clientId: String? = null
        private var clientSecret: String? = null
        private var redirectUri: String? = null
        private var oAuthHost: String = "https://id.kick.com"
        private var authorizationEndpoint: String = "/oauth/authorize"
        private var tokenEndpoint: String = "/oauth/token"
        private var baseUrl: String = "https://api.kick.com/public/v1"
        private var categories: String = "/categories"
        private var categoriesId: String = "/categories/{id}"
        private var tokenIntrospect: String = "/token/introspect"
        private var users: String = "/users"
        private var channels: String = "/channels"
        private var chat: String = "/chat"
        private var moderation: String = "/moderation/bans"
        private var livestreams: String = "/livestreams"
        private var livestreamsStats: String = "/livestreams/stats"
        private var publicKey: String = "/public-key"
        private var events: String = "/events/subscriptions"

        fun tokenStore(tokenStore: RefreshTokenStore) = apply { this.tokenStore = tokenStore }
        fun clientId(clientId: String) = apply { this.clientId = clientId }
        fun clientSecret(clientSecret: String) = apply { this.clientSecret = clientSecret }
        fun redirectUri(redirectUri: String) = apply { this.redirectUri = redirectUri }
        fun oAuthHost(oAuthHost: String) = apply { this.oAuthHost = oAuthHost }
        fun authorizationEndpoint(authorizationEndpoint: String) = apply { this.authorizationEndpoint = authorizationEndpoint }
        fun tokenEndpoint(tokenEndpoint: String) = apply { this.tokenEndpoint = tokenEndpoint }
        fun baseUrl(baseUrl: String) = apply { this.baseUrl = baseUrl }
        fun categories(categories: String) = apply { this.categories = categories }
        fun categoriesId(categoriesId: String) = apply { this.categoriesId = categoriesId }
        fun tokenIntrospect(tokenIntrospect: String) = apply { this.tokenIntrospect = tokenIntrospect }
        fun users(users: String) = apply { this.users = users }
        fun channels(channels: String) = apply { this.channels = channels }
        fun chat(chat: String) = apply { this.chat = chat }
        fun moderation(moderation: String) = apply { this.moderation = moderation }
        fun livestreams(livestreams: String) = apply { this.livestreams = livestreams }
        fun livestreamsStats(livestreamsStats: String) = apply { this.livestreamsStats = livestreamsStats }
        fun publicKey(publicKey: String) = apply { this.publicKey = publicKey }
        fun events(events: String) = apply { this.events = events }

        fun build(): KickConfiguration {
            val tokenStore = requireNotNull(this.tokenStore) { "TokenStore is required" }
            val clientId = requireNotNull(this.clientId) { "ClientId is required" }
            val clientSecret = requireNotNull(this.clientSecret) { "ClientSecret is required" }
            val redirectUri = requireNotNull(this.redirectUri) { "RedirectUri is required" }

            return KickConfiguration(
                tokenStore = tokenStore,
                clientId = clientId,
                clientSecret = clientSecret,
                redirectUri = redirectUri,
                oAuthHost = oAuthHost,
                authorizationEndpoint = authorizationEndpoint,
                tokenEndpoint = tokenEndpoint,
                baseUrl = baseUrl,
                categories = categories,
                categoriesId = categoriesId,
                tokenIntrospect = tokenIntrospect,
                users = users,
                channels = channels,
                chat = chat,
                moderation = moderation,
                livestreams = livestreams,
                livestreamsStats = livestreamsStats,
                publicKey = publicKey,
                events = events,
            )
        }
    }
}
